File: crates/pivot-schema/src/schedules.rs
//! The retransmission schedules the pivot's ladders ride, as DATA
//! (`pivot-schema schedules`): for every [`sip_retransmit::Class`], the wait
//! before each rung the RFC puts on the wire before the class gives up, and
//! the give-up itself, in milliseconds.
//!
//! Exported so a generator that projects a ladder onto a timeline — the cut's
//! drawn-ACK count asks how many copies of a final land before the ACK — reads
//! the one schedule the SUT and the interpreter ride instead of walking T1/T2
//! a second time (ADR-0032 X1). The table is finite because every class gives
//! up, and the walk is [`sip_retransmit::Ladder`]'s own, never restated here.

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use sip_retransmit::{Class, Ladder, Schedule};

/// Every class's schedule, as one exportable document.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct ScheduleTable {
    /// One row per class, in [`Class::ALL`] order.
    pub classes: Vec<ClassSchedule>,
}

/// One class's ladder, walked from the original send to its give-up.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct ClassSchedule {
    /// The class's stable label ([`Class::as_str`]), the same one a counter
    /// names it by.
    pub class: String,
    /// The wait before each rung, rung 1 first: every rung that lands before
    /// the give-up, and no other.
    pub rung_intervals_ms: Vec<u64>,
    /// The bound, measured from the original send: no rung lands at or past it.
    pub give_up_ms: u64,
}

/// The table, one row per class.
pub fn schedule_table() -> ScheduleTable {
    ScheduleTable { classes: Class::ALL.iter().map(|class| row(*class)).collect() }
}

fn row(class: Class) -> ClassSchedule {
    let schedule = Schedule::rfc(class);
    let give_up_ms = millis(schedule.give_up_after());
    let mut rung_intervals_ms = Vec::new();
    if let Some((mut ladder, first)) = Ladder::armed(schedule) {
        rung_intervals_ms.push(millis(first));
        while let Some(wait) = ladder.advance() {
            rung_intervals_ms.push(millis(wait));
        }
    }
    ClassSchedule { class: class.as_str().to_string(), rung_intervals_ms, give_up_ms }
}

fn millis(d: std::time::Duration) -> u64 {
    u64::try_from(d.as_millis()).unwrap_or(u64::MAX)
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Every rung lands before the give-up, and the next would not: the table
    /// is exactly the ladder the class walks, no rung short and none past.
    #[test]
    fn every_row_is_walked_to_its_give_up_and_no_further() {
        let table = schedule_table();
        assert_eq!(table.classes.len(), Class::ALL.len());
        for (row, class) in table.classes.iter().zip(Class::ALL) {
            assert_eq!(row.class, class.as_str());
            let elapsed: u64 = row.rung_intervals_ms.iter().sum();
            assert!(
                elapsed < row.give_up_ms,
                "{}: the last rung lands inside the bound",
                row.class
            );
            let schedule = Schedule::rfc(class);
            let next = u32::try_from(row.rung_intervals_ms.len() + 1).expect("a short ladder");
            let past =
                elapsed + millis(schedule.interval(next).expect("an RFC class always paces"));
            assert!(
                past >= row.give_up_ms,
                "{}: one more rung would land at or past the bound",
                row.class
            );
        }
    }

    #[test]
    fn the_capped_classes_flatten_at_t2_and_the_uncapped_keep_doubling() {
        let table = schedule_table();
        let by = |name: &str| {
            &table.classes.iter().find(|r| r.class == name).expect(name).rung_intervals_ms
        };
        assert_eq!(by("invite-client"), &[500, 1000, 2000, 4000, 8000, 16000]);
        assert_eq!(by("reliable-provisional"), &[500, 1000, 2000, 4000, 8000, 16000]);
        assert_eq!(by("final-2xx"), &[500, 1000, 2000, 4000, 4000, 4000, 4000, 4000, 4000, 4000]);
        assert_eq!(by("non-invite-proceeding"), &[4000; 7]);
    }
}

File: ts/contracts/src/schedules.ts
/**
 * The retransmission schedule table (`pivot-schema schedules`), mirroring
 * `pivot_schema::schedules`'s `ScheduleTable` / `ClassSchedule` — and the
 * table itself, checked in, for a reader with no binary in reach.
 *
 * One schedule, two readers (ADR-0032 X1). `sip-retransmit` owns the RFC 3261
 * §17 / §13.3.1.4 / RFC 3262 §3 ladders; the interpreter paces and counts on
 * it; the cut reads THIS table where it projects a ladder onto a timeline.
 * Nothing here walks T1/T2: {@link SCHEDULES} is data the binary printed, and
 * `@sip/toolchain`'s conformance test holds it byte-equal to what the binary
 * prints now.
 *
 * Each row is one class walked from the original send to its give-up: the
 * wait before rung 1, rung 2, … and no rung that would land at or past the
 * bound — so the list is finite, and a count longer than it is a ladder the
 * RFC has already given up on.
 */
import * as Schema from "effect/Schema"
import { STRICT } from "./strict.js"

/** The stable label `sip_retransmit::Class::as_str` gives each class. */
export const ClassName = Schema.Literals([
  "invite-client",
  "non-invite-client",
  "non-invite-proceeding",
  "cancel-client",
  "invite-server-final",
  "final-2xx",
  "reliable-provisional"
])
export type ClassName = typeof ClassName.Type

/** One class's ladder, walked from the original send to its give-up. */
export const ClassSchedule = Schema.Struct({
  class: ClassName,
  /** The wait before each rung, rung 1 first: every rung landing before the give-up. */
  rung_intervals_ms: Schema.Array(Schema.Int),
  /** The bound from the original send: no rung lands at or past it. */
  give_up_ms: Schema.Int
})
export interface ClassSchedule extends Schema.Schema.Type<typeof ClassSchedule> {}

/** Every class's schedule, one row per class in `Class::ALL` order. */
export const ScheduleTable = Schema.Struct({
  classes: Schema.Array(ClassSchedule)
})
export interface ScheduleTable extends Schema.Schema.Type<typeof ScheduleTable> {}

export const decodeScheduleTable = Schema.decodeUnknownEffect(ScheduleTable, STRICT)
export const decodeScheduleTableSync = Schema.decodeUnknownSync(ScheduleTable, STRICT)

/** The table as `pivot-schema schedules` prints it. */
export const SCHEDULES: ScheduleTable = {
  classes: [
    {
      class: "invite-client",
      give_up_ms: 32000,
      rung_intervals_ms: [500, 1000, 2000, 4000, 8000, 16000]
    },
    {
      class: "non-invite-client",
      give_up_ms: 32000,
      rung_intervals_ms: [500, 1000, 2000, 4000, 4000, 4000, 4000, 4000, 4000, 4000]
    },
    {
      class: "non-invite-proceeding",
      give_up_ms: 32000,
      rung_intervals_ms: [4000, 4000, 4000, 4000, 4000, 4000, 4000]
    },
    {
      class: "cancel-client",
      give_up_ms: 32000,
      rung_intervals_ms: [500, 1000, 2000, 4000, 4000, 4000, 4000, 4000, 4000, 4000]
    },
    {
      class: "invite-server-final",
      give_up_ms: 32000,
      rung_intervals_ms: [500, 1000, 2000, 4000, 4000, 4000, 4000, 4000, 4000, 4000]
    },
    {
      class: "final-2xx",
      give_up_ms: 32000,
      rung_intervals_ms: [500, 1000, 2000, 4000, 4000, 4000, 4000, 4000, 4000, 4000]
    },
    {
      class: "reliable-provisional",
      give_up_ms: 32000,
      rung_intervals_ms: [500, 1000, 2000, 4000, 8000, 16000]
    }
  ]
}

/** The wait before each rung of `cls`'s ladder, rung 1 first, to its give-up. */
export const rungIntervalsMs = (cls: ClassName): ReadonlyArray<number> => {
  const row = SCHEDULES.classes.find((r) => r.class === cls)
  if (row === undefined) throw new Error(`no schedule for class ${cls}`)
  return row.rung_intervals_ms
}

/**
 * The wait before each of a message's `rungs` repeats, rung 1 first: the gaps
 * `stated` where the capture measured them (a step's `retransmit_intervals_ms`,
 * §6.9), otherwise `cls`'s schedule. A count longer than either list repeats
 * the last gap, as `Schedule::exact` does past its list: steady pacing rather
 * than a class the message never chose.
 */
export const rungGapsMs = (
  cls: ClassName,
  rungs: number,
  stated?: ReadonlyArray<number>
): ReadonlyArray<number> => {
  const gaps = stated !== undefined && stated.length > 0 ? stated : rungIntervalsMs(cls)
  return Array.from({ length: rungs }, (_, r) => gaps[Math.min(r, gaps.length - 1)]!)
}
